### DISCUSSION TITLE
Is there any known tooling to inspect event loop states of a running program?

### DISCUSSION THREAD

**Discussion by RaitoBezarius** [capnproto/capnproto#2194] [Q&A]
Hi there, I am working on a project that is transitioning to a kj-based event loop, and we are interested to see if we can obtain some tooling for debuggability of the software, including ways for our users to send us dumps of various event loop states so we can diagnose easily bugs on our side.

If such a tooling does not exist, my understanding is that it should be theoretically possible to construct it by hooking a program running with a kj event loop and knowledge of the various internals, I wonder what the developers who are used to work on the internals think about this idea and whether we could benefit from some guidance if we were to embark on building such a thing.

Thanks!

**Comment by kentonv** [comment]
Some things that are available:
* `Promise<T>::trace()` will return the async trace of what a Promise is waiting on.
* `TaskSet::trace()` will return async traces for all promises in the set.
* `kj::getAsyncTrace()` will return the async trace leading to the currently-executing event.

In all these cases, "async trace" is like a stack trace, but traces a promise chain.

There isn't currently a way to dump everything that's happening on the event loop. Supporting that would, I suppose, require a lot of extra bookkeeping which could be expensive, and it might be difficult to decide how to handle promise/fulfiller pairs.
